utils/estimated_histogram.hh

```diff
@@ -47,13 +47,359 @@
 #include <chrono>
 #include <seastar/core/metrics_types.hh>
 #include <seastar/core/print.hh>
-
-#include <seastar/core/print.hh>
-
 #include "seastarx.hh"
+#include <seastar/core/bitops.hh>
+#include <limits>
 
 namespace utils {
 
+// Returns largest N such that 2^N <= v.
+// Undefined for v == 0.
+inline constexpr size_t pow2_rank(uint64_t v) {
+    return std::numeric_limits<uint64_t>::digits - 1 - count_leading_zeros(v);
+}
+
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+class pseudo_floatingpoint_histogram {
+    std::array<uint64_t, NumBuckets> buckets;
+public:
+    /**
+     * This is a pseudo floating point implementation of an estimated histogram.
+     * When entering a value:
+     * All values lower than the MIN will be included in the first bucket.
+     * All values higher than MAX will be included the last bucket that serves as the
+     * infinity bucket.
+     *
+     * buckets are distributed as pseudo floating point:
+     * The range [MIN, MAX) is split into log2 ranges.
+     *     ranges = log2(max/min)
+     * Each of that ranges is split according to the number of buckets:
+     *     resolution = (NUM_BUCKETS - 1)/ranges
+     *
+     * For example, if the MIN value is 128, the MAX is 1024 and the number of buckets is 13:
+     *
+     * Anything below 128 will be in the bucket 0, anything above 1024 will be in bucket 13.
+     *
+     * the range [128, 1024) will be split into log2(1024/128) = 3:
+     * 128, 256, 512, 1024
+     *
+     * Each range is split into 12/3 = 4.
+     * 128            | 256            | 512            | 1024
+     * 128 160 192 224| 256 320 384 448| 512 640 768 896|
+     *
+     *
+     * Calculating the bucket limit of bucket i:
+     * The range: 2^(i/4)* min
+     * The sub range: i%4 * range/4
+     *
+     * How to find a bucket index for a value.
+     * The bucket index consist of two part:
+     * higher bits are based on log2(value/min)
+     *
+     * lower bits are based on the high 2 MSB (ignoring the leading 1).
+
+     * for example: 330 (101001010)
+     * higher bits: log2(330/128) = 1
+     * low bit MSB: 330 = 01 (the lower two bits out of the upper 3)
+     * So the index: 101 = 5
+     *
+     *
+     * About the min/max and number of buckets.
+     * ========================================
+     *
+     * For MIN and MAX choose numbers that are a power of 2.
+     * The number of buckets will determine the resolution that should also be a power of 2
+     * So the total number of bucket should be log2(MAX/MIN) * Resolution + 1
+     *
+     * Limitation: You must set the MIN value to be higher then the resolution.
+     * For example, for a 2 bits resolution MIN should be 2^2 = 4 or higher.
+     *
+     */
+    using histogram_type = pseudo_floatingpoint_histogram<Min, Max, NumBuckets>;
```

Comment:
Please use just `pseudo_floatingpoint_historgram` to reduce indirection.

Maybe a nicer name is approx_exponential_histogram, no one needs to care about the implementation details.

Reply:
sure